node/src/actors/chain_manager/mining.rs

```diff
@@ -365,6 +377,83 @@ impl ChainManager {
                 dr_state.data_request.collateral
             };
 
+            // Check if we have enough collateralizable unspent outputs before starting
+            // retrieval
+            let block_number_limit = self
+                .chain_state
+                .block_number()
+                .saturating_sub(collateral_age);
+            if !check_commit_collateral(
+                collateral_amount,
+                &mut own_utxos_copy,
+                own_pkh,
+                &self.chain_state.unspent_outputs_pool,
+                timestamp,
+                // The block number must be lower than this limit
+                block_number_limit,
+            ) {
+                log::debug!("Mining data request: Insufficient collateral, the data request need {} mature wits", collateral_amount);
+                continue;
+            }
+
+            // Refrain from trying to resolve any more requests if we have already hit the limit
+            // of retrievals per epoch.
+            let mut start_retrieval_count = cloned_retrieval_count.load(atomic::Ordering::Relaxed);
+            let mut final_retrieval_count =
+                start_retrieval_count.saturating_add(added_retrieval_count);
+
+            if final_retrieval_count > maximum_retrieval_count {
+                log::info!("{} Refrained from resolving data request {} for epoch #{} because it contains {} \
+                        sources, which added to the sources that have already been retrieved ({}) would total {} \
+                        retrievals, which exceed current limit per epoch ({}). This limit exists for performance and \
+                        security reasons. You can increase the limit (AT YOUR OWN RISK) by adjusting the \
+                        `data_request_max_retrievals_per_epoch` inside the `[mining]` section in the `witnet.toml` \
+                        configuration file.",
+                           Yellow.bold().paint("[Mining]"),
+                           Yellow.bold().paint(dr_pointer.to_string()),
+                           Yellow.bold().paint(current_epoch.to_string()),
+                           Yellow.bold().paint(added_retrieval_count.to_string()),
+                           Yellow.bold().paint(start_retrieval_count.to_string()),
+                           Yellow.bold().paint(final_retrieval_count.to_string()),
+                           Yellow.bold().paint(maximum_retrieval_count.to_string())
+                );
+
+                continue;
+            } else {
+                // Update `current_retrieval_count` thanks to interior mutability of the
+                // `cloned_retrieval_count` reference. This is a recursive operation so as
+                // to guarantee addition and prevent potential race conditions in a concurrent
+                // scenario.
+                loop {
+                    let internal_retrieval_count_result = cloned_retrieval_count
+                        .compare_exchange_weak(
```

Comment:
Since some recent versions of Rust we can use

https://doc.rust-lang.org/std/sync/atomic/struct.AtomicU16.html#method.fetch_update

which implements the same logic as this `loop`: try to update an atomic value.

So we can try to refactor this code, and maybe even add some tests.

Reply:
Also I think this limit is no longer correctly checked, it should be applied after checking VRF eligibility.